boards/arm/stm32l4/nucleo-l432kc/src/stm32_dac_wgen.c

```diff
@@ -0,0 +1,284 @@
+
+static struct dac_wgen_s g_dac_wgen =
+{
+  .dac           = NULL,
+  .dac_dmabuffer = stm32l4_dac1_dmabuffer,
+  .samples       = CONFIG_NUCLEOL432KC_DAC_WGEN_SAMPLES,
+  .waveform_freq = ((float)CONFIG_NUCLEOL432KC_DAC_WGEN_FREQ)
+};
+
+/****************************************************************************
+ * Private Function Prototypes
+ ****************************************************************************/
+
+static float waveform_func(float x);
+static void waveform_init(struct dac_wgen_s *dac_wgen, float (*f)(float));
+
+/****************************************************************************
+ * Private Functions
+ ****************************************************************************/
+
+/****************************************************************************
+ * Name: waveform_func
+ *
+ * Description:
+ *   Modulation function. This function must return values from <0.0, 1.0>!
+ *
+ ****************************************************************************/
+
+static float waveform_func(float x)
+{
+  DEBUGASSERT(x >= 0 && x <= 2 * M_PI);
+
+  /* Sine modulation */
+
+  return (sinf(x) + 1.0f) / 2.0f;
+}
+
+/****************************************************************************
+ * Name: waveform_init
+ *
+ * Description:
+ *   Initialize waveform
+ *
+ ****************************************************************************/
+
+static void waveform_init(struct dac_wgen_s *dac_wgen, float (*f)(float))
+{
+  int i;
+  float value;
+
+  /* Get phase step to achieve one sine waveform period */
+
+  dac_wgen->phase_step = (float)(2 * M_PI / dac_wgen->samples);
+
+  /* Initialize sine table */
+
+  for (i = 0; i < dac_wgen->samples; i += 1)
+    {
+      /* We need sine in range from 0 to 1.0 */
+
+      value = (sinf(dac_wgen->phase_step * i) + 1.0f) / 2.0f;
+      dac_wgen->dac_dmabuffer[i] = (uint16_t)(value * 4095);
+    }
+}
+
+/****************************************************************************
+ * Name: dac_wgen_start
+ ****************************************************************************/
+
+static int dac_wgen_start(struct dac_wgen_s *dac_wgen)
+{
+  DAC_ENABLE(dac_wgen->dac, true);
+  DAC_START_DMA(dac_wgen->dac);
+
+  return OK;
+}
+
+/****************************************************************************
+ * Name: dac_wgen_stop
+ ****************************************************************************/
+
+static int dac_wgen_stop(struct dac_wgen_s *dac_wgen)
+{
+  DAC_STOP_DMA(dac_wgen->dac);
+  DAC_ENABLE(dac_wgen->dac, false);
+
+  return OK;
+}
+
+/****************************************************************************
+ * Name: dac_wgen_setup
+ ****************************************************************************/
+
+int dac_wgen_setup(struct dac_wgen_s *dac_wgen)
+{
+  struct dac_dev_s *dac;
+
+  int ret = OK;
+
+  dac = stm32l4_dacinitialize(0);
+  if (dac == NULL)
+    {
+      printf("ERROR: Failed to get DAC interface\n");
```

Comment:
Suggestion: use syslog() instead of printf() so it could be handle better in kernel level (moved to log, etc)